text/0098-generalized-templates.md

```diff
@@ -0,0 +1,277 @@
+
+### Syntax 
+The proposed syntax looks as follows ```template(?bound1: type1, ?bound2: type2) =>``` and types specified must be valid schema types. Schema types are chosen over AST types because they allow us to more precisely specify the type for sets and records. Each new typed slot requires a distinct variable name. However, within the ```when/unless``` clause the same typed slot can be used multiple times. The grammar of template declarations is as follows. ```Type``` is taken directly from the Cedar documentation for schema types. The template syntax (```template(?bound1: type1, ?bound2: type2) =>```) must go after all the declarations. 
+
+```
+Template := 'template' '(' ( TypeAnnotation { ',' TypeAnnotation } ) ')' '=>'
+
+TypeAnnotation := ‘?’ IDENT ‘:’ Type
+
+Type := Path | SetType | RecType
+```
+
+### JSON Policy Format 
+The changes to the JSON policy format would include another attribute ```typed_slots``` and would look as follows: 
+```
+"typed_slots": [
+    { "type": "University::Department", "id": "?department1" }, 
+    { "type": "University::Department", "id": "?department2" }, 
+]
+```
+
+In the JSON policy format, typed slots can be referred to with the same syntax as what is done for ```?principal``` and ```?resource```, except with the identifier being replaced. It would look as follows: 
+```
+"slot": "?department1"
+``` 
+
+
+### ```?principal``` and ```?resource``` Slots
+
+```?principal``` and ```?resource``` can optionally have their types annotated in ```template()```. A similiar effect can be achieved using the ```is``` operator in the condition. However, this would allow templates to be instantiated with slots that always evaluate to false. By having an explicit type annotation we prevent these instantiations from being linked in the first place. 
+
+Aside from the optional type annotation for ```?principal``` and ```?resource``` slots, there are some other differences between them and user defined slots. For one, ```?principal``` and ```?resource``` slots must appear in the scope of the template and can optionally be in the condition of the template, however user defined slots can only appear in the condition clause. Because of this difference ```?principal``` and ```?resource``` slots can only be filled in with entity types and user defined slots can be filled in with any ```SchemaType```.
+
+### What is allowed and not allowed?
+1. Every user defined slot needs to have a type associated with it 
+2. Disallow ```?action``` and ```?context``` slots
+3. Annotated ```?principal``` and ```?resource``` types must be an entity type 
+4. All slots within the template declaration (```template()```) must be used in the template 
```

Comment:
Given this, it'd be nice to somehow "allowlist" the use of `?resource` and `?principal` too somehow.